validation/gas_token.go

```diff
@@ -0,0 +1,104 @@
+package validation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/stretchr/testify/require"
+)
+
+func testGasToken(t *testing.T, chain *superchain.ChainConfig) {
+	l1Client, err := ethclient.Dial(superchain.Superchains[chain.Superchain].Config.L1.PublicRPC)
+	require.NoError(t, err, "Failed to connect to the Ethereum client at RPC url %s", chain.PublicRPC)
+	defer l1Client.Close()
+
+	err = CheckGasToken(chain, l1Client)
+	require.NoError(t, err)
+}
+
+func CheckGasToken(chain *superchain.ChainConfig, l1Client EthClient) error {
+	l2Client, err := ethclient.Dial(chain.PublicRPC)
+	if err != nil {
+		return err
+	}
+	defer l2Client.Close()
+
+	weth9PredeployAddress := superchain.MustHexToAddress("0x4200000000000000000000000000000000000006")
+	want := "0000000000000000000000000000000000000000000000000000000000000020" + // offset
+		"000000000000000000000000000000000000000000000000000000000000000d" + // length
+		"5772617070656420457468657200000000000000000000000000000000000000" // "Wrapped Ether" padded to 32 bytes
+	gotName, err := getHexString("name()", weth9PredeployAddress, l2Client)
+	if err != nil {
+		return err
+	}
+	if want != gotName {
+		return fmt.Errorf("predeploy WETH9.name(): want=%s, got=%s", want, gotName)
+	}
+
+	l1BlockPredeployAddress := superchain.MustHexToAddress("0x4200000000000000000000000000000000000015")
+	isCustomGasToken, err := getBool("isCustomGasToken()", l1BlockPredeployAddress, l2Client)
+	if err != nil && !strings.Contains(err.Error(), "execution reverted") {
+		// Pre: reverting is acceptable
+		return err
+	} else {
+		// Post: must be set to false
+		if isCustomGasToken {
+			return fmt.Errorf("L1Block.isCustomGasToken() must return false")
+		}
+	}
+
+	isCustomGasToken, err = getBool("isCustomGasToken()", superchain.Addresses[chain.ChainID].SystemConfigProxy, l2Client)
+	if err != nil && !strings.Contains(err.Error(), "execution reverted") {
+		// Pre: reverting is acceptable
+		return err
+	} else {
+		// Post: must be set to false
+		if isCustomGasToken {
+			return fmt.Errorf("SystemConfigProxy.isCustomGasToken() must return false")
+		}
+	}
+	return nil
+}
+
+func getBytes(method string, contractAddress superchain.Address, client EthClient) ([]byte, error) {
```

Comment:
Could these functions would become methods on a new client type that we could define? Then they can be reused by the mock and by the real eth client?

For example, we could call it a `HandyClient` and it embeds an `EthClient`, exposing more ergonomic methods to the user. But it's constructor takes the interface so we can inject a mock to make a `FakeHandyClient`.

Reply:
I'm open to that idea, but that seems like follow-up work that could be addressed after the test decoupling when we create the `ValidationSuite` struct